File: code/04-selection/python/select_by_sorting.py
"""SelectBySorting: 정렬 후 rank번째 값을 바로 읽는 baseline(Theta(n log n)).
rank는 0-based(0..n-1), 정렬 장/L03 코드와 같은 색인 관례."""

# snippet:select-by-sorting:start
def select_by_sorting(a, rank):
    return sorted(a)[rank]
# snippet:select-by-sorting:end

if __name__ == "__main__":
    data = [31, 8, 48, 73, 11, 3, 20, 29, 65, 15]
    print("input:", ",".join(str(x) for x in data))
    for rank in (1, 6):
        print("rank:", rank)
        print("result:", select_by_sorting(data, rank))

File: code/04-selection/c/select_by_sorting.c
#include <stdio.h>
#include <stdlib.h>

// snippet:select-by-sorting:start
static int compare_int(const void *pa, const void *pb) {
    int a = *(const int *)pa, b = *(const int *)pb;
    return (a > b) - (a < b);
}

static int select_by_sorting(const int *a, int n, int rank) {
    int *copy = malloc((size_t)n * sizeof *copy);
    for (int i = 0; i < n; i++) copy[i] = a[i];
    qsort(copy, (size_t)n, sizeof *copy, compare_int);
    int result = copy[rank];
    free(copy);
    return result;
}
// snippet:select-by-sorting:end

int main(void) {
    int data[] = {31, 8, 48, 73, 11, 3, 20, 29, 65, 15};
    int n = sizeof(data) / sizeof(data[0]);
    printf("input: ");
    for (int i = 0; i < n; i++) printf("%d%s", data[i], i + 1 < n ? "," : "\n");
    int ranks[] = {1, 6};
    for (int i = 0; i < 2; i++) {
        printf("rank: %d\n", ranks[i]);
        printf("result: %d\n", select_by_sorting(data, n, ranks[i]));
    }
    return 0;
}
